### PR TITLE
Transport abstraction, WebSocket support, bug fixes

### PR DESCRIPTION
**Pull request by djkaty**
This branch abstracts the TCP socket layer from IrcConnection and allows IrcConnection to use pluggable transports instead.

Transports must implement the new IIrcTransportManager. Two transports are provided:

- IrcTcpTransport, providing the same functionality as in the existing implementation
- IrcWebSocketTransport, which allows IRC connections over a WebSocket

The read thread in IrcConnection is moved to IrcTcpTransport. IrcWebSocketTransport does not require a read thread as it is a wrapper around websocket-sharp which is event-driven.

IIrcTransportManager specifies an event-driven receive handler OnMessageReceived which IrcConnection subscribes to. Therefore, the read thread in IrcConnection is replaced by a read queue. The OnMessageReceived handler in IrcConnection (Transport_OnMessageReceived) adds received messages to the read queue.

All transports support AutoRetry and AutoReconnect as these are managed in IrcConnection.

Note that due to the way websocket-sharp is implemented, IrcWebSocketTransport supports AutoRetry but not a multiple server list.

The write buffers, WriteLine etc. are moved into WriteThread which is renamed WriteController to signify it also includes helper methods. This unifies all stream writes to all transports. When WriteController has written a line, it fires a new event OnWriteLine, which IrcConnection subscribes to. The event handler in IrcConnection fires its own OnWriteLine event to end-user applications.

Connect() now supports passing an IIrcTransportManager as the only parameter - pre-populated with connection details - or a transport with a server list or port, plus the original Connect() overloads for backwards compatibility. These will use IrcTcpTransport by default, or whatever the current transport is if it has been changed previously.

The transport in use can be fetched via the new property IrcConnection.Transport. This can be used to modify the connection parameters before calling Connect().

In addition, numerous connection-related bugs are fixed:

- If a new connection is made with a smaller server list, a potential IndexOutOfBoundsException bug is fixed
- AutoRetry would never attempt more than once due to an incorrect comparison
- The number of AutoRetry's is reset when starting a new connection. Before, less than the potential maximum could be attempted (if the comparison operator had been correct).
- AutoRetry recursion has been changed to a do {...} while loop
- The write thread would not reset its queues when reconnecting, meaning that existing queued data would attempt to be written to the new connection incorrectly
- The read thread moved to IrcTcpTransport now cleanly closes the stream when stopping the thread rather than using the evil Thread.Abort().

Other changes:

- ProxyType.cs has been moved to the IrcTransport folder to reflect its proper use.

- The read queue has been updated to use a generic collection Queue<string> instead of a .NET 1.0 non-type-safe collection.

- DCC chat: IrcFeatures.Encoding no longer exists so DccChat.cs wouldn't compile. I have changed it to use the system default encoding (System.Text.Encoding.Default) so that it compiles.

Syntax and testing:

Test.cs has been updated to show all the new possible syntaxes for connecting using standard TCP or WebSockets, using an explicit or implicit transport instance. Please see this for usage details.

I have added a /reconnect command in Test.cs to test reconnection scenarios.

I have tested the library with all of the combinations in Test.cs and can confirm it works as expected. I cannot promise there are no threading bugs whatsoever, especially edge cases relating to disconnects, but I have simulated broken connections or failed connection starts to confirm it works at least superficially.

Breaking changes:

Encoding, UseSsl, ValidateServerCertificate, SslClientCertificate, SocketReceiveTimeout, SocketSendTimeout, ProxyHost, ProxyPort, ProxyType, ProxyUsername and ProxyPassword must now be set on IrcTcpTransport instead of IrcConnection. The property behaviour is otherwise the same (IrcWebSocketTransport uses websocket-sharp's semantics for setting SSL and proxy configurations).

...

Getting this + IRCv3 + Twitch support done is bubbling to the top of things I need to get done for my own project now so I hope we can sort this out quickly. Hope you like the changes!

Katy.

### CODE REVIEW

**Comment by meebey** [comment]
> Showing  8 changed files  with 1,348 additions and 635 deletions.

Wow, you have been very busy :) That will take a bit to review, I know that changing such fundamental code leads to many moved code. It is just hard to review... One thing I noticed quickly is that your coding style (curly braces mostly) does not match the existing code. I will give some pointers as comments in the changes so you know which cases I mean.

**Comment by meebey** [comment]
Also the commits should be squashed and rebased on master, github already says it has conflicts that need to be resolved.

File: examples/test/Test.cs [hunk 1]
```diff
@@ -207,7 +269,14 @@ public static void ReadCommands()
...
                 }
-            } else {
+            } else if (cmd.StartsWith("/reconnect")) {
+                irc.Reconnect();
+
+                // Don't forget to log in again after reconnecting!
+                irc.Login("SmartIRC", "SmartIrc4net Test Bot");
+                irc.RfcJoin("#smartirc-test");
+            }
+            else {
```
Comment (meebey) [inline examples/test/Test.cs]: CS: new extra newline for control structures

**Comment by djkaty** [comment]
Yeah, it's really hard to modify that kind of code without making it all alot worse before it gets better, and trust me, yesterday it was so utterly broken :) I tried to match the coding style, using names like _Foo for private members etc. (which I don't like btw because if their accessibility is protected they're not CLS-compliant), the curly braces are a result of Visual Studio more or less forcing that in auto-formatting. I'm guessing you're mostly referring to the properties. I did notice :)

I don't really know what squash and rebase means, I know what you're saying tho, crunch it into one commit (the first 2 do nothing, the 2nd reverts the first), I'll try and figure out how to do that in team explorer :) It should be able to be auto-merged.

As for review, I'd recommend you look at the new source files and IrcConnection.cs directly rather than the diff because the diff makes little sense with such sweeping changes.

**Comment by meebey** [comment]
Oh for Smuxi I actually specified a Coding Style, see here:
https://github.com/meebey/smuxi/blob/master/HACKING

It uses the some style, but is a good summary of how things should look like

**Comment by meebey** [comment]
We can sort our the squashing later, that is not so important, the diff is already optimizing that away anyhow because the 2nd commit undoes what the 1st commit does, leading to 0 delta :)

**Comment by djkaty** [comment]
You meant I should have "} else {" on the same line right?

Do I need to make a new PR if I modify the coding style or squash?

File: src/IrcFeatures/DccChat.cs [hunk 2]
```diff
@@ -130,8 +130,8 @@ internal override void InitWork(Object stateInfo)
 
             DccChatStartEvent(new DccEventArgs(this));
 
-            _sr = new StreamReader(Connection.GetStream(), Irc.Encoding);
-            _sw = new StreamWriter(Connection.GetStream(), Irc.Encoding);
+            _sr = new StreamReader(Connection.GetStream(), System.Text.Encoding.Default);
+            _sw = new StreamWriter(Connection.GetStream(), System.Text.Encoding.Default);
```
Comment (meebey) [inline src/IrcFeatures/DccChat.cs]: This is definitely wrong since relying on the Default encoding of the OS is wrong in most cases. The library supports setting an explicit encoding as a feature. So the transport needs to expose the used encoding for the IRC connection.

**Comment by meebey** [comment]
The changes in total looks very promising, nicely done!

File: src/IrcFeatures/DccChat.cs (hunk 2 above)
Comment (djkaty) [inline src/IrcFeatures/DccChat.cs]: That was just a hack to make it compile, I don't know how to resolve it. 'Irc' here is an IrcFeatures which has no Encoding property. The IrcConnection is never passed to DccChat. How did you get it to compile?

Comment (djkaty) [inline src/IrcFeatures/DccChat.cs]: DCC chat is a separate socket connection if I remember right, so it doesn't rely on the original connection.

**Comment by meebey** [comment]
If the SmartIrc4net solution has no code formattings defined yet, I should add that. For Smuxi I had already done that.

**Comment by meebey** [comment]
> Do I need to make a new PR if I modify the coding style or squash?

Yeah just make changes, commit and push, you don't need to care about squashing for now. I don't review commits, I review the delta :)

**Comment by djkaty** [comment]
Okay, looked through the coding style, I was mostly on track, I'll go through and make the necessary changes. I'm going to start adding CAP REQ support (IRC 3.2) in another branch later, so let me know when you've figured out everything I've ruined so I can fix it :D Glad you liked it so far :)

**Comment by djkaty** [comment]
Ok, that should have got most of the egregious coding style differences - let me know :)

File: src/IrcFeatures/DccChat.cs (hunk 2 above)
Comment (meebey) [inline src/IrcFeatures/DccChat.cs]: Right, it does not use the original connection, but switching from one specified encoding to an unknown one is unexpected from an application PoV. Say you were talking on #foo in UTF-8 and now some_user sends you ISO-8859-1 via a DDC chat

File: examples/test/Test.cs [hunk 3]
```diff
@@ -130,15 +128,78 @@ public static void Main(string[] args)
         irc.OnError += new ErrorEventHandler(OnError);
         irc.OnRawMessage += new IrcEventHandler(OnRawMessage);
 
-        string[] serverlist;
-        // the server we want to connect to, could be also a simple string
-        serverlist = new string[] {"irc.freenode.org"};
-        int port = 6667;
+        // ===============================================
+        // Standard IRC over TCP
+        // ===============================================
```
Comment (meebey) [inline examples/test/Test.cs]: I would prefer if you can make one project example for regular TCP and one for websockets instead of all the comments :-D

Comment (djkaty) [inline examples/test/Test.cs]: Are you sure? The two examples will be identical for a few lines of connection code.

File: src/IrcFeatures/DccChat.cs (hunk 2 above)
Comment (djkaty) [inline src/IrcFeatures/DccChat.cs]: Sure, I'm not disagreeing with that, my confusion is how the code worked in the first place since Encoding didn't exist on IrcFeatures and the IrcConnection wasn't passed to DccChat.

Shall I just modify the constructor of DccChat (or whatever) to allow it to use the current connection's text encoding?

Comment (meebey) [inline src/IrcFeatures/DccChat.cs]: https://github.com/meebey/SmartIrc4net/blob/master/src/IrcFeatures/DccChat.cs#L63 shows how it worked before. The ctor of DccChat() takes IrcFeatures and that class extends IrcConnection which had Encoding as a property before. Thus I suggested to add Encoding to the Transport interface so it can use Irc.Transport.Encoding here.

File: src/IrcTransport/IrcWebSocketTransport.cs [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+        /// </summary>
+        public event ReadLineEventHandler OnMessageReceived;
+
+#pragma warning disable CS3003 
+        /// <summary>
+        /// The underlying WebSocket of the connection
+        /// </summary>
+        public WebSocket Socket {
+            get; protected set;
+        } = null;
```
Comment (meebey) [inline src/IrcTransport/IrcWebSocketTransport.cs]: What is this supposed to do? The code should stay source compatible with .NET 3.5, not sure if I bumped Smuxi to .NET 4.0 already...

File: src/IrcFeatures/DccChat.cs (hunk 2 above)
Comment (djkaty) [inline src/IrcFeatures/DccChat.cs]: IrcFeatures extends IrcClient in that code...

Comment (djkaty) [inline src/IrcFeatures/DccChat.cs]: Oh for goodness sake Im an idiot... IrcClient extends IrcCommands extends IrcConnection, check... I'll fix it :)

File: src/IrcTransport/IrcWebSocketTransport.cs (hunk 4 above)
Comment (djkaty) [inline src/IrcTransport/IrcWebSocketTransport.cs]: Unnecessary remnant of my code changes, I'll remove the "= null" then it will be C# 2.0-compliant.

**Comment by djkaty** [comment]
Requested changes made, just waiting for an answer on test.cs. Really appreciate you taking the time to go through all this with me :)

**Comment by meebey** [comment]
>  I tried to match the coding style, using names like _Foo for private members etc. (which I don't like btw because if their accessibility is protected they're not CLS-compliant), 

I no longer use _ at all in my current CS, see the Smuxi example, only fields are now f_ prefixed, which is CLS-complaint :)

File: src/IrcConnection/IrcConnection.cs [hunk 5]
```diff
@@ -940,38 +687,29 @@ public void WriteLine(string data)
...
-            return false;
+        /// <summary>
+        /// Received from the transport when a line is successfully received from the connection
+        /// </summary>
+        /// <param name="sender"></param>
+        /// <param name="e"></param>
+        private void Transport_OnMessageReceived(object sender, ReadLineEventArgs e)
+        {
+            // Add it to the read queue
+            // Do we need lock (this) here?
```
Comment (meebey) [inline src/IrcConnection/IrcConnection.cs]: This depends on the thread that raises the event and if _ReadQueue is thread-safe or not

File: src/IrcTransport/IrcTcpTransport.cs [hunk 6]
```diff
@@ -0,0 +1,588 @@
...
+using Starksoft.Net.Proxy;
+
+namespace Meebey.SmartIrc4net
+{
+    /// <summary>
+    /// The default TCP-based IRC protocol with optional proxy and SSL support
+    /// </summary>
+    public class IrcTcpTransport : IIrcTransportManager
+    {
+        private TcpClient _TcpClient;
```
Comment (meebey) [inline src/IrcTransport/IrcTcpTransport.cs]: Please feel free to use auto-properties here, C# 1.0 code makes my eye bleed :-D

File: src/IrcTransport/IrcTcpTransport.cs [hunk 7]
```diff
@@ -0,0 +1,588 @@
...
+        /// Connect to the IRC server via TCP, optionally using SSL and/or a proxy
+        /// </summary>
+        /// <param name="address">Address of the IRC server</param>
+        /// <param name="port">Port of the IRC server</param>
+        /// <exception cref="AuthenticationException">Throws if SSL authentication fails</exception>
+        /// <exception cref="CouldNotConnectException">Throws if the TCP connection failed</exception>
+        public void Connect()
+        {
+            try
+            {
```
Comment (meebey) [inline src/IrcTransport/IrcTcpTransport.cs]: CS: no newline

File: src/IrcTransport/IrcTcpTransport.cs [hunk 8]
```diff
@@ -0,0 +1,588 @@
...
+                _TcpClient = new TcpClient();
+                _TcpClient.NoDelay = true;
+                _TcpClient.Client.SetSocketOption(SocketOptionLevel.Socket, SocketOptionName.KeepAlive, 1);
+                // set timeout, after this the connection will be aborted
+                _TcpClient.ReceiveTimeout = _SocketReceiveTimeout * 1000;
+                _TcpClient.SendTimeout = _SocketSendTimeout * 1000;
+
+                if (_ProxyType != ProxyType.None)
+                {
```
Comment (meebey) [inline src/IrcTransport/IrcTcpTransport.cs]: CS: no newline

File: src/IrcTransport/IrcTcpTransport.cs [hunk 9]
```diff
@@ -0,0 +1,588 @@
...
+                    IProxyClient proxyClient = null;
+                    ProxyClientFactory proxyFactory = new ProxyClientFactory();
+                    // HACK: map our ProxyType to Starksoft's ProxyType
+                    Starksoft.Net.Proxy.ProxyType proxyType =
+                        (Starksoft.Net.Proxy.ProxyType)Enum.Parse(
+                            typeof(ProxyType), _ProxyType.ToString(), true
+                        );
+
+                    if (_ProxyUsername == null && _ProxyPassword == null)
+                    {
```
Comment (meebey) [inline src/IrcTransport/IrcTcpTransport.cs]: CS: no newline

File: src/IrcTransport/IrcTcpTransport.cs [hunk 10]
```diff
@@ -0,0 +1,588 @@
...
+                        proxyClient = proxyFactory.CreateProxyClient(
+                            proxyType
+                        );
+                    }
+                    else
```
Comment (meebey) [inline src/IrcTransport/IrcTcpTransport.cs]: CS: no newlines

File: src/IrcTransport/IrcTcpTransport.cs [hunk 11]
```diff
@@ -0,0 +1,588 @@
...
+                    _TcpClient.Connect(_ProxyHost, _ProxyPort);
+                    proxyClient.TcpClient = _TcpClient;
+                    proxyClient.CreateConnection(_Address, _Port);
+                } else {
+                    _TcpClient.Connect(_Address, _Port);
+                }
+
+                Stream stream = _TcpClient.GetStream();
+                if (_UseSsl)
+                {
```
Comment (meebey) [inline src/IrcTransport/IrcTcpTransport.cs]: CS: no newline

File: src/IrcTransport/IrcTcpTransport.cs [hunk 12]
```diff
@@ -0,0 +1,588 @@
...
+                            certs.Add(_SslClientCertificate);
+                            sslStream.AuthenticateAsClient(_Address, certs,
+                                                            SslProtocols.Default,
+                                                            false);
+                        } else {
+                            sslStream.AuthenticateAsClient(_Address);
+                        }
+                    }
+                    catch (IOException)
+                    {
```
Comment (meebey) [inline src/IrcTransport/IrcTcpTransport.cs]: CS: no newline

File: src/IrcTransport/IrcWebSocketTransport.cs [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+        /// <summary>
+        /// Event which fires when a line of text is received from the connection
+        /// </summary>
+        public event ReadLineEventHandler OnMessageReceived;
+
+        /// <summary>
+        /// The underlying WebSocket of the connection
+        /// </summary>
+        [CLSCompliant(false)]
+        public WebSocket Socket {
```
Comment (meebey) [inline src/IrcTransport/IrcWebSocketTransport.cs]: errr, why isn't it CLSCompliant?

File: src/IrcConnection/IrcConnection.cs [hunk 14]
```diff
@@ -1188,16 +810,27 @@ private class WriteThread
             private int            _BelowMediumThresholdCount = 1;
             private int            _BurstCount;
 
-            public AutoResetEvent QueuedEvent;
+            private AutoResetEvent _QueuedEvent;
```
Comment (meebey) [inline src/IrcConnection/IrcConnection.cs]: This should have been a auto-property but became a public field. In private inner-classes that isn't very harmful (nobody can access it except the outer class), so use the original name and add { get; set; } here

File: src/IrcConnection/IrcConnection.cs [hunk 15]
```diff
@@ -1188,16 +810,27 @@ private class WriteThread
...
+            /// Fires when a line has been successfully sent to the network stream
+            /// </summary>
+            public event WriteLineEventHandler OnWriteLine;
+
+            /// <summary>
+            /// Set up using the specified connection
             /// </summary>
             /// <param name="connection"></param>
-            public WriteThread(IrcConnection connection)
+            public WriteController(IrcConnection connection)
```
Comment (meebey) [inline src/IrcConnection/IrcConnection.cs]: Hm, the WriteThread naming was misleading to you? I don't have better names right now, just MessageSender or SendManager...

File: src/IrcConnection/IrcConnection.cs [hunk 16]
```diff
@@ -1219,18 +860,46 @@ public void Stop()
...
+
+            /// <summary>
+            /// Write a line to the stream if priority is Critical, otherwise queue to one of the write buffers
+            /// </summary>
+            /// <param name="data">String to send</param>
+            /// <param name="priority">Priority level</param>
+            public bool WriteLine(string data, Priority priority)
+            {
+                if (priority == Priority.Critical)
+                {
```
Comment (meebey) [inline src/IrcConnection/IrcConnection.cs]: CS

File: src/IrcConnection/IrcConnection.cs [hunk 17]
```diff
@@ -1254,11 +923,12 @@ private void _Worker()
...
 #endif
+                        // CHECK: shouldn't be necessary any longer
                         // only flag this as connection error if we are not
                         // cleanly disconnecting
-                        if (!_Connection.IsDisconnecting) {
-                            _Connection.IsConnectionError = true;
-                        }
+                        //if (!_Connection.IsDisconnecting) {
+                        //    _Connection.IsConnectionError = true;
+                        //}
```
Comment (meebey) [inline src/IrcConnection/IrcConnection.cs]: Such changes can quickly lead the regressions, I think we will need test-case coverage, else nobody can tell what will happen.

**Comment by meebey** [comment]
Finished initial review

File: src/IrcConnection/IrcConnection.cs (hunk 5 above)
Comment (djkaty) [inline src/IrcConnection/IrcConnection.cs]: Looked over the code. I think I will change it to a ConcurrentQueue.

File: src/IrcTransport/IrcWebSocketTransport.cs (hunk 13 above)
Comment (djkaty) [inline src/IrcTransport/IrcWebSocketTransport.cs]: You will have to ask the guy(s) who wrote websocket-sharp...

File: src/IrcConnection/IrcConnection.cs (hunk 14 above)
Comment (djkaty) [inline src/IrcConnection/IrcConnection.cs]: Why does it need to be a property? It's only used internally.

File: src/IrcConnection/IrcConnection.cs (hunk 15 above)
Comment (djkaty) [inline src/IrcConnection/IrcConnection.cs]: WriteThread was a good name before but I moved a lot of WriteLine-type code and the write queues from IrcConnection to WriteThread so it seemed like a good idea to change the class name to avoid confusion.

File: src/IrcConnection/IrcConnection.cs (hunk 17 above)
Comment (djkaty) [inline src/IrcConnection/IrcConnection.cs]: IsConnectionError doesn't exist in IrcConnection anymore. It's handled by the transports. If we are cleanly disconnecting, the original code doesn't set IsConnectionError and neither does the new transport. If we are not cleanly disconnecting, the transport will have already set Transport.IsConnectionError = true by the time we get here, therefore the code should no longer be necessary. I cannot offer you a 100% cast-iron guarantee but that is how it generally works.

How do you want to implement testing for this?

**Comment by djkaty** [comment]
All requested changes made vis a vis my other comments above.

**Comment by djkaty** [comment]
Just pushed a big commit. Although there may seem like some fluff in there, basically all of the changes to the code-base were driven by the test output. Some things just had to be changed to make everything work.

I forgot to mention in the commit message that I changed the default ping interval and timeout to more sensible defaults.

Please note that the use of async/await and some TPL features in the test code means it has a minimum requirement of .NET 4.5 Framework. If you need it to run on .NET 4.0, I can make some changes. The use of automatic property initializers in some code pushes the minimum compiler requirement to C# 6 (the initializers can easily be changed if that's a problem too).

If you really want me to walk you through the hows and whys and wheres of all the bug fixes I can do that, but some of it is very esoteric.

There are 14 NUnit tests for connections, disconnections and re-connections. Please see the XML documentation in IrcConnectionTests.cs for details of what each test does. The tests all pass on this commit.

image

**Comment by djkaty** [comment]
I have made my final commit to this branch. Please see the commit description for the details.

The final code base includes 34 tests on a wide variety of connection scenarios which I have worked through thoroughly and I believe this code can be considered stable.

Minimum requirement .NET Framework 4.5. I can cut this down to .NET 4.0 if required but I recommend leaving it so that we can implement async/await in the next round of code.

Further optimizations to socket timeouts, transactional operations, burst sending and write operations will be committed in a separate branch that I will clone from this branch.

File: src/IrcConnection/IrcConnection.cs [hunk 18]
```diff
@@ -9,7 +9,8 @@
  *
  * Copyright (c) 2003-2009 Mirco Bauer <meebey@meebey.net> <http://www.meebey.net>
  * Copyright (c) 2008-2009 Thomas Bruderer <apophis@apophis.ch>
- * 
+ * Copyright (c) 2015 Katy Coe <djkaty@start.no> <http://www.djkaty.com>
+*
```
Comment (meebey) [inline src/IrcConnection/IrcConnection.cs]: Missing space

File: src/IrcConnection/IrcConnection.cs [hunk 19]
```diff
@@ -50,44 +45,29 @@ public class IrcConnection
...
         private bool             _AutoReconnect;
-        private Encoding         _Encoding = Encoding.Default;
-        public bool EnableUTF8Recode { get; set; }
-        private int              _SocketReceiveTimeout  = 600;
-        private int              _SocketSendTimeout = 600;
-        private int              _IdleWorkerInterval = 60;
+        private int              _IdleWorkerInterval = 5;
         private int              _PingInterval = 60;
-        private int              _PingTimeout = 300;
+        private int              _PingTimeout = 10;
```
Comment (meebey) [inline src/IrcConnection/IrcConnection.cs]: Why changing the default values? 10 seconds is pretty normal on mobile networks with packet loss

Comment (djkaty) [inline src/IrcConnection/IrcConnection.cs]: Faster disconnect detection for half-open TCP connections.

Why does it need to be 300 seconds if we are sending pings every 60 seconds?

File: src/IrcConnection/IrcConnection.cs [hunk 20]
```diff
@@ -1038,146 +846,14 @@ private void _SimpleParser(object sender, ReadLineEventArgs args)
...
-        }
-
         /// <summary>
-        /// 
+        /// A write thread and helper functions to write data to the connection
         /// </summary>
-        private class WriteThread
-        {
+        private class WriteManager
+        {
```
Comment (meebey) [inline src/IrcConnection/IrcConnection.cs]: CS: trailing whitespace

File: src/IrcConnection/IrcConnection.cs [hunk 21]
```diff
@@ -1208,10 +895,21 @@ public WriteThread(IrcConnection connection)
...
+                _QueuedEvent.Reset();
+
+                ((Queue)_SendBuffer[Priority.High]).Clear();
+                ((Queue)_SendBuffer[Priority.AboveMedium]).Clear();
+                ((Queue)_SendBuffer[Priority.Medium]).Clear();
+                ((Queue)_SendBuffer[Priority.BelowMedium]).Clear();
+                ((Queue)_SendBuffer[Priority.Low]).Clear();
+
+                // Using lock for thread-safe access to _Thread
+                lock (this) {
```
Comment (meebey) [inline src/IrcConnection/IrcConnection.cs]: Why is this lock needed? API docs of Thread class says: "This type is thread safe." I would prefer a local var over locking "this"

File: src/IrcConnection/IrcConnection.cs [hunk 22]
```diff
@@ -1222,18 +920,51 @@ public void Stop()
...
-                // the thread will think the connection is broken!
-                _Thread.Join();
-                
-                try {
-                    _Connection._Writer.Close();
-                } catch (ObjectDisposedException) {
+                _QueuedEvent.Set();
+
+                // We have to check _Thread here in case the connection is disconnected before the thread starts
+                lock (this) {
```
Comment (meebey) [inline src/IrcConnection/IrcConnection.cs]: local var instead of lock?

File: src/IrcConnection/IrcConnection.cs [hunk 23]
```diff
@@ -1470,50 +1206,81 @@ public void Start()
...
+                            // Is it time to send a new ping?
+                            if (now.Subtract(_Connection._LastPongReceived).TotalSeconds >= _Connection.PingInterval) {
+                                _Connection.WriteLine(Rfc2812.Ping(pingAddress), Priority.Critical);
+                                _Connection._LastPingSent = now;
+                            }
+                        }
+
+                        // Ping timeout
+                        if (timeout) {
+                            System.Diagnostics.Debug.WriteLine("Ping timeout");
```
Comment (meebey) [inline src/IrcConnection/IrcConnection.cs]: Please use the logger class

File: src/IrcTransport/IrcTcpTransport.cs [hunk 24]
```diff
@@ -0,0 +1,523 @@
...
+        /// <summary>
+        /// The encoding to use to write to and read from the socket.
+        ///
+        /// If EnableUTF8Recode is true, reading and writing will always happen
+        /// using UTF-8; this encoding is only used to decode incoming messages
+        /// that cannot be successfully decoded using UTF-8.
+        ///
+        /// Default: encoding of the system
+        /// </summary>
+        public Encoding Encoding { get; set; } = Encoding.Default;
```
Comment (meebey) [inline src/IrcTransport/IrcTcpTransport.cs]: Use ctor to initialize auto properties

File: src/IrcTransport/IrcTcpTransport.cs [hunk 25]
```diff
@@ -0,0 +1,523 @@
...
+                    lock (_Writer) {
+                        _Writer.Write(data + "\r\n");
+                        _Writer.Flush();
+                    }
+
+                } catch (IOException ex) {
+#if LOG4NET
+                    Logger.Socket.Warn("sending data failed, connection lost");
+#endif
+                    System.Diagnostics.Debug.WriteLine("Sending data failed: " + ex.Message);
```
Comment (meebey) [inline src/IrcTransport/IrcTcpTransport.cs]: Logger

File: src/IrcTransport/IrcWebSocketTransport.cs [hunk 26]
```diff
@@ -0,0 +1,348 @@
...
+        /// <summary>
+        /// Event which fires when a connection error occurs
+        /// </summary>
+        public event Action OnConnectionError;
+
+        /// <summary>
+        /// The underlying WebSocket of the connection
+        /// </summary>
+        [CLSCompliant(false)]
+        public WebSocket Socket { get; set; }
```
Comment (meebey) [inline src/IrcTransport/IrcWebSocketTransport.cs]: Do we need to expose the underlying socket here?

File: src/IrcConnection/IrcConnection.cs (hunk 20 above)
Comment (djkaty) [inline src/IrcConnection/IrcConnection.cs]: Fixed in pending commit

File: src/IrcConnection/IrcConnection.cs (hunk 21 above)
Comment (djkaty) [inline src/IrcConnection/IrcConnection.cs]: In a previous version there was a small chance of deadlock - Thread is thread-safe but not atomic (transactional). Anyhow I re-ran the unit tests and everything is ok so I've removed the locks.

File: src/IrcConnection/IrcConnection.cs (hunk 22 above)
Comment (djkaty) [inline src/IrcConnection/IrcConnection.cs]: Removed, see previous comment.

File: src/IrcConnection/IrcConnection.cs (hunk 23 above)
Comment (djkaty) [inline src/IrcConnection/IrcConnection.cs]: Fixed in IrcConnection and IrcTcpTransport. Where does the logger output go by default?

File: src/IrcTransport/IrcTcpTransport.cs (hunk 24 above)
Comment (djkaty) [inline src/IrcTransport/IrcTcpTransport.cs]: Is this necessary? All mono versions from 3.8 upwards support C# 6 auto-property initializers. Awaiting feedback. I changed .travis.yml elsewhere to use this as it was using mono 2.2 or 2.8 or something like that.

File: src/IrcTransport/IrcTcpTransport.cs (hunk 25 above)
Comment (djkaty) [inline src/IrcTransport/IrcTcpTransport.cs]: Fixed

File: src/IrcTransport/IrcWebSocketTransport.cs (hunk 26 above)
Comment (djkaty) [inline src/IrcTransport/IrcWebSocketTransport.cs]: Originally yes, so the user could alter the WebSocket configuration if required. Now all of the properties are set in IrcWebSocketTransport via proxy properties. Do you want me to change it to private?

File: src/IrcTransport/IrcTcpTransport.cs (hunk 24 above)
Comment (knocte) [inline src/IrcTransport/IrcTcpTransport.cs]: Sorry to jump in, but even if mono >= 3.8 supports auto-property initializers, SmartIrc4net currently requires mono 1.2.6 only, so people with mono < 3.8 would get compiler errors.

I guess you would need to propose a separate PR that raises this dependency check in configure.ac before using this feature here.

Comment (djkaty) [inline src/IrcTransport/IrcTcpTransport.cs]: Np on jumping in but I was told .NET 4.0 was ok :)

Comment (knocte) [inline src/IrcTransport/IrcTcpTransport.cs]: Well, maybe it was a mistake to tell you that, or it's a mistake that the `configure.ac` still requires a Mono version so old that cannot compile the .NET 4.0 profile.

Comment (knocte) [inline src/IrcTransport/IrcTcpTransport.cs]: Besides, I believe C#6 is not included in .NET4.0, maybe we're talking about .NET4.5 or higher.

**Comment by txdv** [comment]
You didn't merge this @meebey ?

**Comment by djkaty** [comment]
https://github.com/djkaty/SmartIrc4net/commits/master

Last year I posted .NET 4.0 support, transport abstraction, WebSockets support, IRCv3 caps support with IDN / Punycode parsing, write queue optimizations and a fairly comprehensive burst sending control, as I had to use SmartIrc4Net in a large distributed bot project. None of the changes were ever merged and I'm working on another project now so I had to clear my Pull requests log. Feel free to look at/use the commits to the repo above if you want these features.

**Comment by txdv** [comment]
Good thing open source lets you fork easily. I thought @meebey would be more thrilled about people working on smartirc4net

**Comment by knocte** [comment]
I guess meebey didn't merge because he never had time to review a monster-commit. Open source etiquette tells you it's better to do small commits.

**Comment by meebey** [comment]
@djkaty @txdv @knocte, yeah these changes are still on my todo list. I have spent hours on reviewing already but since there are so many changes it needs further reviewing and testing.
